### ISSUE TITLE
AbstractBuildMojo does not honour pom dependency exclusions

### ISSUE THREAD

**Issue by manstis** [Vertispan/j2clmavenplugin#72]
If you include a dependency into a J2CL project then exclusions added to the projects `pom.xml` are not honoured.

[Here](https://github.com/manstis/j2cl-exclude-deps/tree/main) is a contrived Maven project that creates a simple POJO.

The project has a dependency (obviously not needed in this example, but there to demonstrate the issue).

[Here](https://github.com/manstis/j2cl-exclude-deps-consumer) is an equally contrived project that consumes the first as a dependency and excludes the transitive dependency.

The `mvn dependency:tree` result shows the transitive dependency is not considered part of the projects dependencies:

```
[INFO] --- maven-dependency-plugin:2.8:tree (default-cli) @ j2cl-exclude-deps-consumer ---
[INFO] org.manstis:j2cl-exclude-deps-consumer:jar:1.0
[INFO] \- org.manstis:j2cl-exclude-deps:jar:1.0:compile
```
This project fails to compile with:
```
[INFO] ---------------< org.manstis:j2cl-exclude-deps-consumer >---------------
[INFO] Building j2cl-exclude-deps-consumer 1.0
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ j2cl-exclude-deps-consumer ---
[INFO] Deleting /home/manstis/workspaces/git/manstis/j2cl-exclude-deps-consumer/target
[INFO] 
[INFO] --- j2cl-maven-plugin:0.17-SNAPSHOT:build (build-js) @ j2cl-exclude-deps-consumer ---
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  1.011 s
[INFO] Finished at: 2021-07-19T13:02:51+01:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal com.vertispan.j2cl:j2cl-maven-plugin:0.17-SNAPSHOT:build (build-js) on project j2cl-exclude-deps-consumer: Failed to build project structure: Some problems were encountered while processing the POMs:
[ERROR] [ERROR] 'distributionManagement.status' must not be specified. @ line 56, column 13: 1 problem was encountered while building the effective model for xerces:xercesImpl:2.8.1
[ERROR] [ERROR] 'distributionManagement.status' must not be specified. @ line 56, column 13
```
I understand the distribution management issue relates to the inclusion of xerces as a transitive dependency.

The problem appears to be that `AbstractBuildMojo.loadDependenciesIntoCache(..)` does not honour exclusions.

It can be [noted](https://github.com/Vertispan/j2clmavenplugin/blob/master/src/main/java/net/cardosi/mojo/AbstractBuildMojo.java#L140) `List<Artifact> dependencies = currentProject.getArtifacts()` returns the full dependency tree.

Use of the deprecated `List<Artifact> dependencies = mavenProject.getDependencyArtifacts()` only includes direct dependencies. It shows `Artifact.getDependencyFilter()` is populated with the dependency filter declared in the `pom.xml`. Unfortunately the collection of artifacts returned with `mavenProject.getArtifacts()` does not have the dependency filter property populated.

I appreciate use `mavenProject.getDependencyArtifacts()` is not acceptable for the J2CL plugin (that needs to transpile the whole _used_ dependency tree including transitive dependencies) but it might help point to a solution.

**Comment by manstis** [comment]
Full build log.

[log.txt](https://github.com/Vertispan/j2clmavenplugin/files/6840634/log.txt)

Notably:

```
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  1.068 s
[INFO] Finished at: 2021-07-19T13:09:29+01:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal com.vertispan.j2cl:j2cl-maven-plugin:0.17-SNAPSHOT:build (build-js) on project j2cl-exclude-deps-consumer: Failed to build project structure: Some problems were encountered while processing the POMs:
[ERROR] [ERROR] 'distributionManagement.status' must not be specified. @ line 56, column 13: 1 problem was encountered while building the effective model for xerces:xercesImpl:2.8.1
[ERROR] [ERROR] 'distributionManagement.status' must not be specified. @ line 56, column 13
[ERROR] -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal com.vertispan.j2cl:j2cl-maven-plugin:0.17-SNAPSHOT:build (build-js) on project j2cl-exclude-deps-consumer: Failed to build project structure
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:215)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:564)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
Caused by: org.apache.maven.plugin.MojoExecutionException: Failed to build project structure
    at net.cardosi.mojo.BuildMojo.execute (BuildMojo.java:228)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:564)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
Caused by: org.apache.maven.project.ProjectBuildingException: Some problems were encountered while processing the POMs:
[ERROR] 'distributionManagement.status' must not be specified. @ line 56, column 13

    at org.apache.maven.project.DefaultProjectBuilder.build (DefaultProjectBuilder.java:207)
    at org.apache.maven.project.DefaultProjectBuilder.build (DefaultProjectBuilder.java:342)
    at net.cardosi.mojo.AbstractBuildMojo.loadDependenciesIntoCache (AbstractBuildMojo.java:194)
    at net.cardosi.mojo.AbstractBuildMojo.loadDependenciesIntoCache (AbstractBuildMojo.java:203)
    at net.cardosi.mojo.AbstractBuildMojo.loadDependenciesIntoCache (AbstractBuildMojo.java:203)
    at net.cardosi.mojo.BuildMojo.execute (BuildMojo.java:220)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:564)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
Caused by: org.apache.maven.model.building.ModelBuildingException: 1 problem was encountered while building the effective model for xerces:xercesImpl:2.8.1
[ERROR] 'distributionManagement.status' must not be specified. @ line 56, column 13

    at org.apache.maven.model.building.DefaultModelProblemCollector.newModelBuildingException (DefaultModelProblemCollector.java:197)
    at org.apache.maven.model.building.DefaultModelBuilder.build (DefaultModelBuilder.java:498)
    at org.apache.maven.model.building.DefaultModelBuilder.build (DefaultModelBuilder.java:440)
    at org.apache.maven.model.building.DefaultModelBuilder.build (DefaultModelBuilder.java:430)
    at org.apache.maven.project.DefaultProjectBuilder.build (DefaultProjectBuilder.java:173)
    at org.apache.maven.project.DefaultProjectBuilder.build (DefaultProjectBuilder.java:342)
    at net.cardosi.mojo.AbstractBuildMojo.loadDependenciesIntoCache (AbstractBuildMojo.java:194)
    at net.cardosi.mojo.AbstractBuildMojo.loadDependenciesIntoCache (AbstractBuildMojo.java:203)
    at net.cardosi.mojo.AbstractBuildMojo.loadDependenciesIntoCache (AbstractBuildMojo.java:203)
    at net.cardosi.mojo.BuildMojo.execute (BuildMojo.java:220)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:564)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
```

**Comment by niloc132** [comment]
Thanks for the report, and sorry for the delay - I've just metged a completely rebuilt version of the plugin that likely has the same "bug", though I am now trying to consider a possible fix. The problem is that your sample doesn't actually make sense (though it does exhibit the bug of course) as a real use case, so it is hard to model correctly and decide if this would be fixed or not. I'm pretty sure it wouldn't solve most cases, but I can't be sure about this one without more details... here goes:

For a single project, ignoring tests, there are two classpaths that are built in maven - the compile classpath, and the runtime classpath. The compile classpath of course is used to compile the source files present in the current project, and the runtime classpath contains the elements necessary to run these source files. The vast majority of the time, these classpaths are the same - a dependency which is listed on one is also on the other (note that Gradle and others can get a good deal more clever here by allowing the project model to declare which ones must be used by downstream compile vs runtime classpaths, but this oversimplification is sufficient for this discussion), but there are a few cases where you would specify scope=provided or scope=runtime to force maven to only add a dependency in one or the other case. As you correctly observed, `MavenProject.getArtifacts()` of `j2cl-exclude-deps-consumer` does not return the `ant` dependency, it is correctly excluded here, since when maven computes the dependencies of that one project, it excludes that transitive dependency.

This is obviously not true if you try to build the `j2cl-exclude-deps` project itself - it lists ant, and maven, not knowing why you listed that, correctly adds `ant` and `ant-launcher` to the compile and runtime classpaths, which was what you requested.

So next, consider what work the j2cl-maven-plugin needs to do in order to create the desired compiled JS at the end of the build (see https://github.com/Vertispan/j2clmavenplugin/#build-process for the in-depth version). 
 * To begin with, the plugin considers not the compiletime classpath, but the runtime one, since Closure needs to link the required JS together and optimize it for runtime. For `j2cl-exclude-deps-consumer`, this does not include `ant`, it was excluded, and the plugin respected that.
 * So, what JS needs to be linked? That runtime classpath consists of two elements, the J2CL output of `j2cl-exclude-deps-consumer` and `j2cl-exclude-deps`. So, we need to transpile some Java to JS for both of those.
 * Pausing for a moment to consider how J2CL does its job - much like javac, it takes a set of plain source files, and a (compile-time) classpath of .jar files containing bytecode. The set of sources files is expected to be small for it to do its job quickly, and the results can be cached if the inputs don't change. The contents of the classpath must be the bytecode that matches the runtime JS "classpath", so that J2CL emits class/method/field references that are consistent even across many separate invocations of J2CL.
 * Okay, so back to running j2cl on these two projects. First, we run on `j2cl-exclude-deps-consumer`, so we compile its sources against its _compile-time_ classpath (we've only considered the runtime classpath up until now). Again, as expected, this only contains `j2cl-exclude-deps` itself, `ant` was correctly excluded. So, we now we also need the bytecode for `j2cl-exclude-deps`.
 * The two steps remaining on our list take the same requirements, so I'll just consider them once - we need the transpiled JS and bytecode for `j2cl-exclude-deps`, the first to be fed to closure-compiler, and the second to be used to correctly transpile `j2cl-exclude-deps-consumer`. So, we need the compile-time classpath of `j2cl-exclude-deps`. This is where your bug strikes - nothing at all to do with your excludes statement, since the `j2cl-exclude-deps` pom has no such excludes. We must follow the pom of `j2cl-exclude-deps` and expect that it has what we need, and whether you call `MavenProject.getArtifacts()` or `MavenProject.getDependencies()`, you get `ant` in that list (note that `MavenProject.getDependencyArtifacts()` returns null here, but as a direct dependency of `j2cl-exclude-deps`, if it were populated, it would also contain `ant`). This is the real "bug".
 * It may seem confusing that we need the bytecode for `j2cl-exclude-deps` rather than just using the bytecode already in the jar. This does make sense though when you consider the preprocessor step that strips out `@GwtIncompatible` members and types. It is possible that in the future this will not be a separate step, but while this simplifies our bytecode step, it doesn't help with j2cl needing the compile-time classpath of `j2cl-exclude-deps`, which still includes `ant`.

So really, there is a second "tree" of dependencies to consider, which is actually arguably not a tree at all, but a graph, and isn't your project's dependencies, but describes the more abstract "things that are needed to perform all the work that must be done to run J2CL and Closure and produce JS". Some options that seem workable, but most have problems of one kind or another:
 * Just skip elements that fail - these errors almost always point to something that doesn't make sense anyway, and silently ignoring errors is a good way to really confuse users later when it would be incorrect to skip a real dependency that was deployed corrupted in some way.
 * Transitively apply exclusions. This is basically what GWT2 does, but GWT2 also allows you to mix and match client and non-client code, and exclude files through configuration, approaches which the authors of j2cl have expressly rejected (their answer would be "the build tool should do the rejection in whatever way it supports it", and the way that maven supports it is to make another artifact, or add some extra build steps to unpack a dependency, modify it, and republish it?). GWT2 achieves this by making a single "whole-world" classpath, rather than compiling each dependency in isolation, which is why you can have cycles (!!!) in dependencies between gwt modules, even across jars. A maven plugin could exist to do this, but it would lose its ability to cache results, you would effectively need a full compile or a much more clever way to detect changes than what we've come up with.
 * Let the plugin guess that the dependency is actually not necessary. This is the case in your example! `j2cl-exclude-deps` actually has no runtime _or_ compiletime dependency on `ant` - a better example would have ant be necessary to compile but not run like an annotation jar or annotation-processor (and be marked with scope=provided), or be necessary to run, but not compile like a "service loader" entry for logging or something else (and be marked with scope=runtime). Unfortunately, just because something doesnt fail to compile when a dependency was removed automatically, doesn't mean it didn't break something. 

This last one does lead me to my only working idea: Lazily build the MavenProject for any dependency - it shouldn't be enough to be included on some project's classpath, something should request an actual task be run on it (like javac or j2cl) to force it to build a full project model. Since `ant` never made it to `j2cl-exclude-deps-consumer`'s runtime classpath, j2cl was not requested to run on it and we wouldn't fail there. But `j2cl-exclude-deps`'s j2cl task still needs accurate gwt-compatible bytecode - if your real project was excluding `ant` and providing a different runtime implementation that match the same API, we can see we would still need it to match, know how to call the right methods, and so `ant`'s source needs to be unpacked, stripped of `@GwtIncompatible`, recompiled with javac, and put on the classpath for `j2cl-exclude-deps`, so we _still_ need that project model.

If we further add a check for "if this is an external project" (so the contents can't change in `j2cl:watch`), and "if this project either has no sources or those sources never had any `@GwtIncompatible` annotations", we could skip referencing the project's dependencies, and so if we were lazily building the project model we could stop there. This adds up to a lot of complex changes though, for a dubious use case.

Concrete suggestions for your project (aside from "give me a real world example so I can help"):
 * Repackage the dependency that is broken, so j2cl-m-p can give it a quick look and say "yep, looks boring, I'll put it on the classpaths required and then ignore it". If making an annotation-processor or the like, this is almost certainly the correct answer.
 * Remove it, since it is apparently unused (if not unused, please update the question so it has that detail)
 * Specify `<dependencyReplacements>`, telling the plugin to rewrite _any_ dependency _across the entire graph_ that matches. Take great care here, if you replace the default values here, you almost certainly need to restore the jsinterop-base entry, since Google doesn't ship a maven dependency of jsinterop-base that is j2cl compatible. If removing `ant` in this way works, it probably should never have been a dependency to begin with? (PRE-SEND-EDIT: Excluding xercesImpl might solve this better? ant depends on xercesImpl as a runtime dependency)
 * We're also considering a `<pomReplacements>` like the above, see #43. This would let you rewrite the `j2cl-exclude-deps` pom only so that it never had `ant` to begin with, but will probably break for the same reason if the `ant` dependency actually belonged there.

As far as I've seen, `xercesImpl` always breaks, but similarly, it is always red flag that something else shouldn't be on the classpath but is. One more takeaway, rather than the possible steps proposed above: add xercesImpl to the default dependency replacements list. We already log those replacements, so if someone has a downstream bug, we will see that xerces was involved and that something else might be wrong.

Fixes to consider for the project:

- [ ] Add xerces:xercesImpl to the default dependencyReplacements list
- [ ] Project implementation could offer a lazy lookup. This could be messier than it looks since more than one "branch" of the graph might need to be updated when the same entries are resolved, so the lookup will need to be sure to close over the rest of the graph and be thread safe.
- [ ] Let the default "stripped-bytecode" task (presently a class called JavacTask) detect that the "stripped-sources" task made no changes, and emit the same bytecode as the "bytecode" task. In cases where sources were not mapped (i.e. they belong to a dependency outside the current project) it seems like we should be able to do this and not bother with the usual approach of listening for its dependencies to change. The hard part is in letting one task detect details, not just output, of another task - by design this isn't something that tasks should be able to observe, so as to correctly cache outputs based on inputs (yet another case where "use gradle" would get us the right answer instantly, with the huge downside of making everyone ... use gradle).

**Comment by manstis** [comment]
Hi @niloc132 thank-you for your detailed reply.

Simplified reproducers can invariably make the reported scenario seem urealistic.

My _real_ projects are here:

- https://github.com/manstis/pmml-codegen/tree/j2clmavenplugin-72
- https://github.com/manstis/marshaller/tree/j2clmavenplugin-72

`pmml-codegen` uses JAXB to generate annotated POJOs from an XSD
It itself is spit into two sub-modules; one to generate interfaces and another to generate concrete POJOs.

`marshaller` is the J2CL project that re-uses the generated POJOs (and hence has a dependency on `pmml-codegen`).
It generates marshallers for the PMML POJOs and then passes the whole lot through the J2CL plugin.

The projects themselves suffer other issues I am working on.. but they compile far enough to exhibit the issue reported here.

I suspect the crux is that the JAXB codegen brings in a compile time dependency that causes the J2CL plugin to fail.

My workaround is to paste the JAXB generated code into the J2CL project.

**Comment by niloc132** [comment]
Thanks for the added detail, I'll run it when I get the chance, but can you clarify how jaxb is actually used? For example, is it used in an annotation processor, or is it being actually compiled to JS? I would assume from "to generate interfaces and ... concrete POJOs" that it is just for codegen? If so, this is probably from an annotation processor , and you can consider shading the dependencies that are needed into the processor - this has the added benefit that any conflicts between your codegen dependencies and any other processors used by downstream projects can not result in runtime errors, since they would already be resolved through packaging.

Can you confirm that adding xerces:xercesImpl as a dependencyReplacement (remember to at least re-add jsinterop-base, and possibly gwt-user and gwt-dev depending on your other dependencies) fixes this issue, or possibly points out a new issue to be concerned with?

**Comment by manstis** [comment]
Hi @niloc132; firstly the quick answer: I am using one of the (many) JAXB Maven Plugins in the `pmml-codegen` project to generate POJOs from an XSD (that is really using XJC under the hood). The POJOs include JAXB annotations. We are then using an annotation processor (in the `marshaller` project) that generates code to unmarshaller/deserialise XML based on the JAXB annotations in the POJOs. Once all of the codegen is complete all dependencies we had on JAXB related libraries disappear completely.

I'll investigate the other options you suggest as I am not familiar with `dependencyReplacement` at the moment.

**Comment by manstis** [comment]
OK, so removing dependency `xerces:xercesImpl` with the following `pom` configuration opens a can of worms.

`pom.xml`
```
          <dependencyReplacements>
            <dependencyReplacement>
              <original>xerces:xercesImpl</original>
            </dependencyReplacement>
            <dependencyReplacement>
              <original>com.google.jsinterop:base</original>
              <replacement>com.vertispan.jsinterop:base:1.0.0-SNAPSHOT</replacement>
            </dependencyReplacement>
            <dependencyReplacement>
              <original>org.realityforge.com.google.jsinterop:base</original>
              <replacement>com.vertispan.jsinterop:base:1.0.0-SNAPSHOT</replacement>
            </dependencyReplacement>
            <dependencyReplacement>
              <original>com.google.gwt:gwt-user</original>
            </dependencyReplacement>
            <dependencyReplacement>
              <original>com.google.gwt:gwt-dev</original>
            </dependencyReplacement>
            <dependencyReplacement>
              <original>com.google.gwt:gwt-servlet</original>
            </dependencyReplacement>
          </dependencyReplacements>
```
`Error`
```
[INFO] --- j2cl-maven-plugin:0.17-SNAPSHOT:build (build-js) @ marshaller-api ---
Removing dependency com.google.jsinterop:base:jar:1.0.0:compile, replacing with com.vertispan.jsinterop:base:jar:1.0.0-SNAPSHOT
Removing dependency com.google.jsinterop:base:jar:1.0.0:compile, replacing with com.vertispan.jsinterop:base:jar:1.0.0-SNAPSHOT
Removing dependency com.google.jsinterop:base:jar:1.0.0:compile, replacing with com.vertispan.jsinterop:base:jar:1.0.0-SNAPSHOT
Removing dependency com.google.jsinterop:base:jar:1.0.0:compile, replacing with com.vertispan.jsinterop:base:jar:1.0.0-SNAPSHOT
Removing dependency xerces:xercesImpl:jar:2.8.1:runtime, no replacement
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  04:56 min
[INFO] Finished at: 2021-08-02T16:49:21+01:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal com.vertispan.j2cl:j2cl-maven-plugin:0.17-SNAPSHOT:build (build-js) on project marshaller-api: Failed to build project structure: 1 problem was encountered while building the effective model for com.sun.xml.bind.external:rngom:2.3.0
[ERROR] [FATAL] Non-resolvable parent POM for com.sun.xml.bind.mvn:jaxb-parent:2.3.0: Could not transfer artifact com.sun.xml.bind:jaxb-bom-ext:pom:2.3.0 from/to maven-default-http-blocker (http://0.0.0.0/): Blocked mirror for repositories: [releases.java.net (http://maven.java.net/content/repositories/releases/, default, releases+snapshots), shapshots.java.net (http://maven.java.net/content/repositories/snapshots/, default, releases+snapshots), jvnet-nexus-staging (http://maven.java.net/content/repositories/staging/, default, releases+snapshots), netbeans (http://bits.netbeans.org/nexus/content/groups/netbeans, default, releases)] @ com.sun.xml.bind.mvn:jaxb-parent:2.3.0, /home/manstis/.m2/repository/com/sun/xml/bind/mvn/jaxb-parent/2.3.0/jaxb-parent-2.3.0.pom, line 48, column 13
[ERROR]  for project com.sun.xml.bind.external:rngom:2.3.0
[ERROR] -> [Help 1]
``` 
I'd like your comments before I proceed down the rabbit hole.

**Comment by niloc132** [comment]
So it looks like you will need to prune more dependencies in this way, and that seems like a deep rabbit hole to wander down.

My advice is to try shading the apt jar you're making, so that you don't have random dependencies like this, or keep playing whack-a-mole with broken poms in central.

**Comment by niloc132** [comment]
To add, the other ideas I proposed _could_ work, given a) that they are implementable (they aren't easy, but they might be possible), and b) that none of the problem poms are actually required in order to run the code you are trying to write.

**Comment by manstis** [comment]
@niloc132 

> Can you confirm that adding xerces:xercesImpl as a dependencyReplacement (remember to at least re-add jsinterop-base, and possibly gwt-user and gwt-dev depending on your other dependencies) fixes this issue, or possibly points out a new issue to be concerned with?

I've pushed a new commit to the consuming project including all the `dependencyReplacements` needed.. however it results in infinite recursion within the J2CL MOJO and a stack overflow. Feel free to have a look.

To be honest use of `dependencyReplacement` with an `original` and no `replacement` seems to be trying to do exactly what Maven's `dependency\exclusions` does but is needed for every transient dependency and not just the top level ones. For now, my simple work-around, is to copy the generated POJOs to the `marshaller` project.

**Comment by niloc132** [comment]
No, dependencyReplacement is not at all what excludes does - see my first reply, excludes still works exactly as you expect _for a single project_. Remember, for a traditional JVM-based project, the bytecode is uploaded in jars to maven central already - but not only is there no central repository or format for J2CL-transpiled code, but the ABI isn't fixed and the j2cl project could change any part of the format at any time - an entire J2CL project should always be built using the same version of J2CL across all dependencies/modules. This means that you must recompile all dependencies as part of your own build, and each dependency must be built using its own classpath, not your final project classpath.

--

That is a gigantic list of replacements - if you are removing this many things, your dependency tree isn't actually intended to be used in a J2CL project. I strongly urge you to repackage dependencies so they make sense for J2CL (for APT, this probably means shading down to a standalone jar, which should be a straightforward process, especially if you never intend your processor to be used by anyone but you. If you do intend others to use your processor and you have this many dependencies, you should definitely shade it and relocate classes, to protect their relative classpaths from one another).

--

I can't build pmml-codegen from a clean start:
```
[ERROR] Failed to execute goal on project pmml-codegen-api: Could not resolve dependencies for project org.manstis:pmml-codegen-api:jar:0.0.0: Could not find artifact org.jvnet.jaxb2_commons:jaxb2-basics:jar:0.12.1-SNAPSHOT in vertispan-j2cl-snapshots (https://repo.vertispan.com/j2cl/) -> [Help 1]
```

Which snapshot repo should also be listed in https://github.com/manstis/pmml-codegen/blob/j2clmavenplugin-72/pom.xml to allow me to resolve this?

**Comment by manstis** [comment]
1) Yes, thank-you for reminding me... I understand the need to generate the byte code for all dependencies.

2) Yes, indeed and it has a nasty (code) smell about it. There's no need to have the code generator exposed to J2CL.

3) Are you sure you're using the `j2clmavenplugin-72` branch? The [version](https://github.com/manstis/pmml-codegen/blob/j2clmavenplugin-72/pom.xml#L38) there is `0.12.0` readily available in [Maven Central](https://search.maven.org/artifact/org.jvnet.jaxb2_commons/jaxb2-basics/0.12.0/jar).

That said, since the issue is the inclusion of the code generator dependency in the "model JAR" it probably makes more sense to re-package that to just contain the POJOs and minimal dependencies (for the annotations).

**Comment by manstis** [comment]
Hi @niloc132 I have elected to build a shaded JAR containing the minimal generated source code; and consequentially remove the compile time dependency on the JAXB/XJC annotation processor thus removing the same from the J2CL compilation. It seems to work. We may forever disagree about whether it's reasonable to expect a pom's dependency exclusions to apply to the J2CL plugin however I understand your comments and respect your leadership on this plugin, J2CL and "GWT's" future.

**Comment by niloc132** [comment]
@manstis thanks for the update - I do not think you are unreasonable about this, but I also do not think you understand the specific case and how excludes work with regard to how even maven itself for normal projects functions. The important thing to remember is that maven expects your dependencies to _already be compiled_ - the jars are available in some repo with compiled code. 

We can't do that for j2cl though, so we have to build those artifacts ourselves. This "build the things that for normal maven/jvm would already be built" is the _only_ part of the build where we ignore excludes, and this accurately reflects how those artifacts in maven were built - if you depend on ant, you can't assert that you only download ant bytecode which was built without certain dependencies.

Certainly I will grant that _how_ we have to resolve those dependencies is different from a hypothetical case where you recursively checked out and built each of your dependencies, but those differences are small and hard to reconcile - and only apply in odd cases like this one, where some project has an entirely unnecessary dependency, or in the specific case of annotation processors.
